### PR TITLE
New Plugin Event

### PR DESCRIPTION
**Pull request by wilsonge**
Pull Request for Issue #10351 
#### Summary of Changes

This succeeds @rdeutz 's PR at #10351 - it adds a generic on before validation event to allow manipulation of data to validate including the com_users PR that he mentioned earlier
#### Testing Instructions

Ensure there are no PHP errors in any core components

### CODE REVIEW

File: libraries/legacy/model/admin.php [hunk 1]
```diff
@@ -163,7 +156,8 @@ public function __construct($config = array())
 			array(
 				'delete'       => 'content',
 				'save'         => 'content',
-				'change_state' => 'content'
+				'change_state' => 'content',
+				'validate'     => 'content'
```
Comment (zero-24) [inline libraries/legacy/model/admin.php]: Can we habe a , here too?

File: libraries/legacy/model/form.php [hunk 2]
```diff
@@ -28,6 +28,34 @@
...
+	 */
+	protected $events_map = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param   array  $config  An optional associative array of configuration settings.
+	 *
+	 * @see     JModelLegacy
+	 * @since   12.2
```
Comment (zero-24) [inline libraries/legacy/model/form.php]: Should it not since 3.6/3.7?

**Comment by wilsonge** [comment]
@zero-24 fixed

**Comment by zero-24** [comment]
Thanks i have just noticed that there is a issue with the tests?!

```
There were 6 errors:
1) JModelFormTest::testValidateReturnsFalseOnJformException
JDatabaseExceptionExecuting: No database selected SQL=SELECT id, rules
FROM `#__viewlevels`
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver/mysqli.php:660
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver.php:1494
/home/travis/build/joomla/joomla-cms/libraries/joomla/access/access.php:935
/home/travis/build/joomla/joomla-cms/libraries/joomla/user/user.php:453
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:296
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:163
/home/travis/build/joomla/joomla-cms/libraries/legacy/model/form.php:333
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/legacy/model/JModelFormTest.php:57
2) JModelFormTest::testValidateReturnsFalseOnJformFalse
JDatabaseExceptionExecuting: No database selected SQL=SELECT id, rules
FROM `#__viewlevels`
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver/mysqli.php:660
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver.php:1494
/home/travis/build/joomla/joomla-cms/libraries/joomla/access/access.php:935
/home/travis/build/joomla/joomla-cms/libraries/joomla/user/user.php:453
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:296
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:163
/home/travis/build/joomla/joomla-cms/libraries/legacy/model/form.php:333
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/legacy/model/JModelFormTest.php:101
3) JModelFormTest::testValidateModifiesTagsArray with data set #0 (array('1'), array('1'))
JDatabaseExceptionExecuting: No database selected SQL=SELECT id, rules
FROM `#__viewlevels`
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver/mysqli.php:660
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver.php:1494
/home/travis/build/joomla/joomla-cms/libraries/joomla/access/access.php:935
/home/travis/build/joomla/joomla-cms/libraries/joomla/user/user.php:453
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:296
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:163
/home/travis/build/joomla/joomla-cms/libraries/legacy/model/form.php:333
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/legacy/model/JModelFormTest.php:150
4) JModelFormTest::testValidateModifiesTagsArray with data set #1 (array(array('1')), array(array('1'), '1'))
JDatabaseExceptionExecuting: No database selected SQL=SELECT id, rules
FROM `#__viewlevels`
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver/mysqli.php:660
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver.php:1494
/home/travis/build/joomla/joomla-cms/libraries/joomla/access/access.php:935
/home/travis/build/joomla/joomla-cms/libraries/joomla/user/user.php:453
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:296
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:163
/home/travis/build/joomla/joomla-cms/libraries/legacy/model/form.php:333
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/legacy/model/JModelFormTest.php:150
5) JModelFormTest::testValidateModifiesTagsArray with data set #2 (array('1', array('2')), array('1', array('2')))
JDatabaseExceptionExecuting: No database selected SQL=SELECT id, rules
FROM `#__viewlevels`
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver/mysqli.php:660
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver.php:1494
/home/travis/build/joomla/joomla-cms/libraries/joomla/access/access.php:935
/home/travis/build/joomla/joomla-cms/libraries/joomla/user/user.php:453
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:296
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:163
/home/travis/build/joomla/joomla-cms/libraries/legacy/model/form.php:333
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/legacy/model/JModelFormTest.php:150
6) JModelFormTest::testValidatePassesDataAndGroupToJform
JDatabaseExceptionExecuting: No database selected SQL=SELECT id, rules
FROM `#__viewlevels`
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver/mysqli.php:660
/home/travis/build/joomla/joomla-cms/libraries/joomla/database/driver.php:1494
/home/travis/build/joomla/joomla-cms/libraries/joomla/access/access.php:935
/home/travis/build/joomla/joomla-cms/libraries/joomla/user/user.php:453
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:296
/home/travis/build/joomla/joomla-cms/libraries/cms/plugin/helper.php:163
/home/travis/build/joomla/joomla-cms/libraries/legacy/model/form.php:333
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/legacy/model/JModelFormTest.php:186

```

**Comment by rdeutz** [comment]
I have tested this item :white_check_mark: successfully on 162e40f05afe9997b5b7028874562593b012eb5c<br /><br />works as expected (with the simplify user plugin) when also applying #10350 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10751">issues.joomla.org/joomla-cms/10751</a>.</sub>

**Pull request merged**